### ISSUE TITLE
Generic parameter referencing a nested class

### ISSUE THREAD

**Issue by lanzz** [microsoft/pyright#8176]
**Describe the bug**
When a type parameter on a generic class forward-references a nested class within the generic class itself, Pyright will emit a "Class definition depends on itself" error and will reveal the bound parameter as "Unknown". For the same test code, `mypy` correctly determines the parameter binding and reveals the expected nested class. The actual code works fine at runtime, the class definition isn't really depending on itself and compiles with no issues.

**Code or Screenshots**
```python
from typing import Generic, TypeVar

T = TypeVar('T')

class Base(Generic[T]):

    def __init__(self, o: T) -> None:
        self.o = o

    def get(self) -> T:
        return self.o

class NotNested:
    pass

class A(Base[NotNested]):
    pass

class B(Base['B.Nested']):

    class Nested:
        pass

reveal_type(A(NotNested()).get)
reveal_type(B(B.Nested()).get)
```

Output from Pyright:
```
.../test.py
  .../test.py:24:17 - error: Class definition for "B" depends on itself (reportGeneralTypeIssues)
  .../test.py:30:13 - information: Type of "A(NotNested()).get" is "() -> NotNested"
  .../test.py:31:13 - information: Type of "B(B.Nested()).get" is "() -> Unknown"
1 error, 0 warnings, 2 informations 
```

Reference output from `mypy`:
```
test.py:30: note: Revealed type is "def () -> test.NotNested"
test.py:31: note: Revealed type is "def () -> test.B.Nested"
Success: no issues found in 1 source file
```

**VS Code extension or command-line**
Initially encountered in VS Code extension, but reproducible from command-line with Pyright 1.1.367 (latest on PyPI as of time of posting this).

**Comment by erictraut** [comment]
As indicated by the error message, this is a circular reference, and it's not supported in pyright. For details, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=circular-references).

In this particular case, mypy's lack of an error here is arguably a bug. An attribute access expression `B.Nested` requires that `Nested` be bound to `B`. To perform this binding, the type of `B` must be known — including its class hierarchy. That means `B` must be evaluated before `B.Nested` can be evaluated.

Also, `B` in the expression `B.Nested` refers to the post-decorated class `B` (if there were a class decorator present), and applying a decorator to a class requires that you know its type — including its type hierarchy. 

I don't know enough about your use case, but there's almost always a straightforward way to remove circular references like this.

In any case, I don't consider this a bug. Pyright is working as intended here.

**Comment by lanzz** [comment]
An _attribute access_ at the starting point of the `B` class declaration, sure - but this isn't an attribute access. Arguably when PEP 649 gets implemented in 3.14, this perceived circularity should completely evaporate, since the annotations (and therefore the type of the class) should be evaluated after the entire class is defined, and not at arbitrary points during its definition.
